plutus-wallet-api/ledger/Ledger/Typed/Tx.hs

```diff
@@ -0,0 +1,419 @@
+
+-- | A 'TxOut' tagged by a phantom type: and the connection type of the output.
+newtype TypedScriptTxOut a = TypedScriptTxOut { unTypedScriptTxOut :: TxOut }
+-- | Create a 'TypedScriptTxOut' from a correctly-typed data script, an address, and a value.
+makeTypedScriptTxOut
+    :: forall out
+    . ScriptInstance out
+    -> CompiledCode (DataType out)
+    -> Value.Value
+    -> TypedScriptTxOut out
+makeTypedScriptTxOut ct d value =
+    let outTy = PayToScript (DataScript (fromCompiledCode d))
+    in TypedScriptTxOut @out $ TxOutOf (scriptAddress ct) value outTy
+
+-- | A 'TxOutRef' tagged by a phantom type: and the connection type of the output.
+newtype TypedScriptTxOutRef a = TypedScriptTxOutRef { unTypedScriptTxOutRef :: TxOutRef }
+
+-- | A public-key 'TxOut'. We need this to be sure that it is not a script output.
+newtype PubKeyTxOut = PubKeyTxOut { unPubKeyTxOut :: TxOut }
+-- | Create a 'PubKeyTxOut'.
+makePubKeyTxOut :: Value.Value -> PubKey -> PubKeyTxOut
+makePubKeyTxOut value pubKey = PubKeyTxOut $ pubKeyTxOut value pubKey
+
+-- | A typed transaction, tagged by two phantom types: a list of connection types for the outputs,
+-- and a list of connection types for the inputs. The script outputs and inputs must have the correct
+-- corresponding types.
+data TypedTx (ins :: [Type]) (outs :: [Type]) = TypedTx {
+    tyTxTypedTxIns   :: HListF (TypedScriptTxIn outs) ins,
+    tyTxPubKeyTxIns  :: [PubKeyTxIn],
+    tyTxTypedTxOuts  :: HListF TypedScriptTxOut outs,
+    tyTxPubKeyTxOuts :: [PubKeyTxOut],
+    tyTxForge        :: !Value.Value,
+    tyTxValidRange   :: !SlotRange
+    }
+
+baseTx :: TypedTx '[] '[]
+baseTx = TypedTx {
+    tyTxTypedTxIns = HNilF,
+    tyTxPubKeyTxIns = [],
+    tyTxTypedTxOuts = HNilF,
+    tyTxPubKeyTxOuts = [],
+    tyTxForge = Value.zero,
+    tyTxValidRange = Interval.always
+    }
+
+-- | Adds a 'TypedScriptTxOut' to a 'TypedTx'.
+--
+-- Adding a new typed transaction output is only safe if there are no typed transaction
+-- inputs yet. Otherwise those inputs would need to change to take the new data script as
+-- an argument.
+addTypedTxOut
+    :: forall outs newOut
+    . TypedScriptTxOut newOut
+    -> TypedTx '[] outs
+    -> TypedTx '[] (newOut ': outs)
+-- We're changing the type so we can't use record update syntax :'(
+addTypedTxOut out TypedTx {
+    tyTxTypedTxOuts,
+    tyTxPubKeyTxOuts,
+    tyTxTypedTxIns=HNilF,
+    tyTxPubKeyTxIns,
+    tyTxForge,
+    tyTxValidRange } = TypedTx {
+      tyTxTypedTxOuts=HConsF out tyTxTypedTxOuts,
+      tyTxPubKeyTxOuts,
+      tyTxTypedTxIns=HNilF,
+      tyTxPubKeyTxIns,
+      tyTxForge,
+      tyTxValidRange }
+
+-- | Adds a 'TypedScriptTxIn' to a 'TypedTx'.
+addTypedTxIn
+    :: forall ins outs newIn
+    . TypedScriptTxIn outs newIn
+    -> TypedTx ins outs
+    -> TypedTx (newIn ': ins) outs
+-- We're changing the type so we can't use record update syntax :'(
```

Comment:
Maaaybe. I started looking at this before and decided I should not waste lots of time doing clever things with lens before I finished the rest of the interface. I'll have another look.

Reply:
Ah yes, it's a bit tricky:
- Lens can't autogenerate a type-changing lens for the outputs because the type variable appears in the inputs too.
- I can write a limited type-changing lens by hand, but it would have to be written just like this function, so I'm not sure how much we gain.